AppEngine Standard Application Version resource.

khanali21 opened the pull request:
<!-- 
Note: You may see "This branch is out-of-date with the base branch"
when you submit a pull request. This is fine! We don't use the GitHub
merge button to merge PRs, and you can safely ignore that message.

Thanks for contributing!
-->
This commit adds the AppEngine standard application version resource. 
This is work in progress, however I would like to seek initial review from the contributors.

modular-magician commented:
Hello!  I am a robot who works on Magic Modules PRs.

I have detected that you are a community contributor, so your PR will be assigned to someone with a commit-bit on this repo for initial review.  They will authorize it to run through our CI pipeline, which will generate downstream PRs.

Thanks for your contribution!  A human will be with you soon.

modular-magician commented:
Hi!  I'm the modular magician, I work on Magic Modules.
This PR seems not to have generated downstream PRs before, as of fae29170cb50ebdc56dd238d8b34578f7bd1f295. 
## Pull request statuses
No diff detected in Ansible.
No diff detected in Inspec.
## New Pull Requests
I built this PR into one or more new PRs on other repositories, and when those are closed, this PR will also be merged and closed.
depends: https://github.com/terraform-providers/terraform-provider-google-beta/pull/817 
depends: https://github.com/GoogleCloudPlatform/terraform-google-conversion/pull/98 
depends: https://github.com/terraform-providers/terraform-provider-google/pull/3799

khanali21 commented:
Is this conflict build/terraform-beta expected?

rileykarson commented on products/appengine/api.yaml:
```diff
@@ -63,3 +63,537 @@ objects:
+      - !ruby/object:Api::Type::Array
+        name: 'zones'
+        item_type: Api::Type::String
+        description: 'The Google Compute Engine zones that are supported by this version in the App Engine flexible environment.'
```
For descriptions, I'd suggest always using the `description: |` form. We treat them as preformatted markdown, while this treats them as a string w/ the newline removed.

abombss commented on products/appengine/api.yaml:
```diff
@@ -63,3 +63,524 @@ objects:
           - ALLOW
           - DENY
+  - !ruby/object:Api::Resource
+    name: 'StandardAppVersion'
```
@rileykarson Its interesting that some fields are only applicable to Flex and some for Standard.  In this case the models are not totally divergent but different enough that it potentially makes management under a single resource model challenging.  It looks like the Azure folks recently made the shift to split divergent resources into separate models for Windows vs Linux vm's. terraform-providers/terraform-provider-azurerm#2807

rileykarson replied:
Yep! We've used some split resources in the Google provider as well, such as https://www.terraform.io/docs/providers/google/r/compute_ssl_certificate.html and https://www.terraform.io/docs/providers/google/r/compute_managed_ssl_certificate.html which are the same resource API-side.

I'm comfortable with either of a single resource or a split one. We'll need to be mindful of `flex`-only fields and eliminating `env` as a user-configurable field  (https://github.com/GoogleCloudPlatform/magic-modules/pull/1884/files#diff-da914abaf076a320ef68301fe61306cfR186) if we're splitting the resource.

khanali21 commented on products/appengine/api.yaml:
```diff
@@ -63,3 +63,524 @@ objects:
+    description: |
+      A single firewall rule that is evaluated against incoming traffic
+      and provides an action to take on matched requests.
+    base_url: 'apps/{{project}}/versions'
```
Thank you very much @rileykarson  I will be committing changes shortly.

khanali21 replied:
@rileykarson Need your help in something which I am struggling with.  I did try to use
the base_url path as apps/{{project}}/services/{{service}}/versions
but then I start getting 500 internal server error as the POST does not contain the name of the service e.g. "default"
/v1/apps/myapp/services//versions?alt=json

If I try to introduce a property called service then it fails with error: 
invalid or unknown key: service


```
resource "google_app_engine_standard_app_version" "myapp" {
  version_id = "v31"
  runtime = "nodejs10"
  api_config = {}
  endpoints_api_service = {}
  entrypoint {
    shell = "node index.js"
  }
```

your help is appreciated.

khanali21 replied:
Actually that was the original reason why I had to use the base_url as apps/{{project}}/versions
as this was the only way I could put the whole path inside the project parameter. But obviously that was wrong.
Conceptually I am struggling with the understanding of base_url and its connection with path parameters. How does the path parameters take the values. I see that "project" can come from provider , what about other parameters in the path? How they can be defined?
I did try adding the property service and setting 
```
exclude: true
```
in the terraform.yaml but doesn't seem to work either.

khanali21 replied:
I think I got it using 
```
- !ruby/object:Api::Type::ResourceRef
        name: 'service'
        resource: 'service'
        imports: 'name'
        description: |
          Service
        required: true  

```

khanali21 replied:
I am still struggling with this though moving forward.
So now I can reference the service from the Service resource. But I get the error:
"message": "Invalid JSON payload received. Unknown name \"service\" at 'version': Cannot find field.",

Looks like what I want to do is to exclude the "service" property from the Version but still have it in my parameters for the base_url.
I have been looking into examples from compute resource. 
Need some guidance on the yaml schema for the api.yml
and also understand the difference between, 
input: true
exclude: true
is_set: true

Your help is really appreciated.

khanali21 replied:
ignore above too, found url_param_only flag to not include in the body.
Missed it twice on reading. :)

rileykarson replied:
It looks like this part is still possibly unresolved:

> what about other parameters in the path? How they can be defined?

The `replaceVars` function we defined in the Terraform provider will pull the value from these sources, in order:

1. For special vars (`project`, `region`, `zone`) .tf config, then the provider config
2. .tf config (most fields are inferred here)
3. provider config (`config.go`)

Since `service` wasn't a property in Terraform previously, it couldn't be inferred from user config. Most APIs will silently ignore values that aren't really part of the request body, but this one doesn't. In those cases, yep, `url_param_only` is useful :+1:

khanali21 commented on products/appengine/terraform.yaml:
```diff
@@ -24,6 +24,21 @@ overrides: !ruby/object:Overrides::ResourceOverrides
         test_env_vars:
           org_id: :ORG_ID
+  StandardAppVersion: !ruby/object:Overrides::Terraform::ResourceOverride
+    id_format: "{{project}}/{{serviceId}}/{{name}}"
```
right, I was confused about that.

khanali21 commented:
Hi, 
Why the concourse-ci/status is not getting reported? Is there anything I need to do? Please advise.
Thanks in advance.

rileykarson commented:
Ah, triggering Concourse builds is a manual process for external contributions, once a member of the team has looked over the code briefly. I didn't notice this getting pushed, it didn't trigger a notification for some reason. If you push and want me to take a look, include a comment as well! Those should always result in a notication.

It looks like there's merge conflicts in the `build/` folders. I started a build, but I _expect_ they'll cause it to fail. Can you rebase on top of `master` and update those folders so there's no conflict?

rileykarson commented:
Looks like the build failed because of Ansible, which has a required field for supported resources. Can you add an exclusion for this resource in `appengine/ansible.yaml`? Similar to https://github.com/GoogleCloudPlatform/magic-modules/blob/master/products/compute/ansible.yaml#L62-L63

rileykarson commented:
Oops, sorry for the extra round trip- can you do the same for `Service` as well?

khanali21 commented:
No my fault should have checked the api.yml for any other resource. Many thanks for your support @rileykarson

modular-magician commented:
Hi!  I'm the modular magician, I work on Magic Modules.
I see that this PR has already had some downstream PRs generated. Any open downstreams are already updated to your most recent commit, 3565fbb9ee26c525d1bf0ab910a23e17bd20187e. 
## Pull request statuses
terraform-provider-google-beta already has an open PR.
terraform-google-conversion already has an open PR.
terraform-provider-google already has an open PR.
No diff detected in Ansible.
No diff detected in Inspec.
## New Pull Requests
I didn't open any new pull requests because of this PR.

khanali21 commented on templates/terraform/examples/app_engine_standard_app_version.tf.erb:
```diff
@@ -0,0 +1,21 @@
+
+resource "google_app_engine_app_version" "<%= ctx[:primary_resource_id] %>" {
+  version_id = "v31"
+  project = "${google_app_engine_application.app.project}/services/default"
```
I agree with you, in fact I had started only with files initially and then added Zip and other fields. I would push the changes as suggested..

modular-magician commented:
Hi!  I'm the modular magician, I work on Magic Modules.
I see that this PR has already had some downstream PRs generated. Any open downstreams are already updated to your most recent commit, c8f287762a10f3384e85f8e993dc61b506d9074a. 
## Pull request statuses
terraform-provider-google-beta already has an open PR.
terraform-google-conversion already has an open PR.
terraform-provider-google already has an open PR.
No diff detected in Ansible.
No diff detected in Inspec.
## New Pull Requests
I didn't open any new pull requests because of this PR.

rileykarson commented on templates/terraform/examples/app_engine_standard_app_version.tf.erb:
```diff
@@ -0,0 +1,22 @@
+  endpoints_api_service = {}
+  entrypoint {
+    shell = "node index.js"
+  }
```
```suggestion
  }
}
```

khanali21 replied:
Thank you very much Riley for your support. I have made the changes. Can you also tell me if I can run those tests locally to validate. I did do compile and test the newly built provider locally, but I don't think I am running the go tests for the provider.

rileykarson replied:
Instructions for running the provider tests are at https://github.com/terraform-providers/terraform-provider-google/blob/master/.github/CONTRIBUTING.md#tests, they'll be ran directly from the provider directory.

Make sure to include a pattern, probably `TestAccAppEngineStandardAppVersion_`. All of our tests are acceptance tests, and running other product suites wouldn't be useful.

khanali21 replied:
Thanks for the response. I will do.
Also one more question. When we release this new resource for App Version, would it be compatible with Terraform 0.11? Or would I need 0.12+?

rileykarson replied:
It'll be compatible with both!

modular-magician commented:
Hi!  I'm the modular magician, I work on Magic Modules.
I see that this PR has already had some downstream PRs generated. Any open downstreams are already updated to your most recent commit, a0031834cdfe34a1d686c2b0f923c105e56b1c3b. 
## Pull request statuses
terraform-provider-google-beta already has an open PR.
terraform-google-conversion already has an open PR.
terraform-provider-google already has an open PR.
No diff detected in Ansible.
No diff detected in Inspec.
## New Pull Requests
I didn't open any new pull requests because of this PR.

khanali21 commented:
@rileykarson Hi Riley, Is there anything I need to do? Please let me know.

modular-magician commented:
Hi!  I'm the modular magician, I work on Magic Modules.
I see that this PR has already had some downstream PRs generated. Any open downstreams are already updated to your most recent commit, 278b48a39b7aa3d627a7a63335e9d4ec1d810f15. 
## Pull request statuses
terraform-provider-google-beta already has an open PR.
terraform-google-conversion already has an open PR.
terraform-provider-google already has an open PR.
No diff detected in Ansible.
No diff detected in Inspec.
## New Pull Requests
I didn't open any new pull requests because of this PR.

rileykarson commented on templates/terraform/examples/app_engine_standard_app_version.tf.erb:
```diff
@@ -0,0 +1,23 @@
+}
+
+
+resource "google_app_engine_app_version" "<%= ctx[:primary_resource_id] %>" {
```
```suggestion
resource "google_app_engine_standard_app_version" "<%= ctx[:primary_resource_id] %>" {
```

rileykarson commented on templates/terraform/examples/app_engine_standard_app_version.tf.erb:
```diff
@@ -0,0 +1,23 @@
+  project = "${google_app_engine_application.app.project}"
+  service = "default"
+  runtime = "nodejs10"
+  api_config = {}
```
Make sure that fields that aren't being added right now aren't included, they'll cause the generated test to fail.

rileykarson commented on templates/terraform/examples/app_engine_standard_app_version.tf.erb:
```diff
@@ -0,0 +1,23 @@
+  runtime = "nodejs10"
+  api_config = {}
+  endpoints_api_service = {}
+  entrypoint {
```
Can you test adding files with `deployment` and using `handlers`?

modular-magician commented:
Hi!  I'm the modular magician, I work on Magic Modules.
I see that this PR has already had some downstream PRs generated. Any open downstreams are already updated to your most recent commit, 3166665e614896dee97ad6546a86249037d7570d. 
## Pull request statuses
terraform-provider-google-beta already has an open PR.
terraform-google-conversion already has an open PR.
terraform-provider-google already has an open PR.
No diff detected in Ansible.
No diff detected in Inspec.
## New Pull Requests
I didn't open any new pull requests because of this PR.

rileykarson commented:
I'm assuming the CLA change is from performing an action in the GitHub web UI, which can have a different associated email than you use locally on your machine. Do you mind performing a rebase there to squash your commits down and fix the email?

khanali21 commented:
Hi @rileykarson , apologies I was out of office, so I have taken care of the changes you suggested, and I have also added the zip file deployment as this helps in the use case I am working for terraform.

Help needed:
I have just started looking at the test cases generated as part of the CI process. I am running into an issue on executing the test when I want to update the existing version resource. I changed the update_url with verb PATCH, however I am not sure how to use the custom field value for update_mask. 
So I need the following url
uri: https://appengine.googleapis.com/v1/apps/sample/services/default?alt=json&migrateTraffic=False&updateMask=split
method: PATCH
 

but when I run the terraform apply with existing resource:
 PATCH /v1/apps/sample/services/default?alt=json&updateMask=id%2Cruntime%2CenvVariables%2Cdeployment%2Centrypoint


I think it uses the id, runtime, envVariables, deployment, entrypoint fields by default.

rileykarson commented:
I'm not sure what you mean by `split`, `updateMask` values should be set for every modified field based on the following: https://github.com/terraform-providers/terraform-provider-google-beta/pull/817/files#diff-be7a1b17548460da485ebadb23552368R376-R418

modular-magician commented:
Hi!  I'm the modular magician, I work on Magic Modules.
I see that this PR has already had some downstream PRs generated. Any open downstreams are already updated to your most recent commit, 418b57854405733c43afbc3d69a903884123bc5a. 
## Pull request statuses
terraform-provider-google-beta already has an open PR.
terraform-google-conversion already has an open PR.
terraform-provider-google already has an open PR.
No diff detected in Ansible.
No diff detected in Inspec.
## New Pull Requests
I didn't open any new pull requests because of this PR.

khanali21 commented:
@rileykarson Just updated the PR with the changes. Here is the summary of the changes:
1. Added the remaining fields of handler and have tested it with the following:

```
resource "google_app_engine_standard_app_version" "myapp" {
  version_id = "v2"
  service = "default"
  runtime = "nodejs10"
  entrypoint {
    shell = "node app.js"
  }
  handlers {
    security_level = "SECURE_OPTIONAL"
    static_files {
      path = "stylesheets/"
      upload_path_regex = "stylesheets/.*"
    }
    url_regex = "/stylesheets/(.*)"
  }
  deployment {
    zip {
      source_url = "https://storage.googleapis.com/test/hello-world.zip"
    }
    
  } 
  env_variables = {
    port = "8080"
  } 
  
}

```

2. I have also added the async operation check for creation and deletion. Both have been tested by me.

khanali21 commented:
@rileykarson There are two  issues I want to resolve:

1. Right now I can update the existing version resource. I get a 404 error. I am trying investigate this.
2. I need to make the version_id as the primary resource id , so that if I change the version_id it should delete the old version and create a new version, somewhat similar to google_storage_bucket where if you change the name of the bucket , the old bucket gets deleted and a new bucket is created.

Your help is truly appreciated.

khanali21 commented:
@rileykarson Hi Riley, can you please advise? Is there anything I thing I need to do, I have tested it locally. I have also rebased it to the latest master.

rileykarson commented on third_party/terraform/website/docs/r/app_engine_standard_app_version.html.markdown:
```diff
@@ -0,0 +1,72 @@
+---
```
This file should be generated automatically, you can remove it.

khanali21 replied:
Thanks for your support, I have made the require changes. 
But I think I need to upload the hello-world.zip in the test-fixtures folder, is it possible to do that.
hello-world.zip is based on the https://github.com/GoogleCloudPlatform/nodejs-docs-samples/tree/master/appengine/hello-world/standard

I have tested by deploying using the new resource.

once again thanks for your support.

khanali21 replied:
I have one more question, what are the guidelines for implementing the resource update . So essentially the patch verb for updating the service uses fields which are not the same as create version. The patch verb actually updates the version for splitting the traffic.

Right now I am able to create and delete the version using the new resource.
But in the context of appengine version resource (https://cloud.google.com/appengine/docs/admin-api/reference/rest/v1/apps.services.versions/patch),
the patch should split the traffic between multiple versions. 

And I am trying to figure out what is the right way to implement it. Should I create a separate resource for splitting the traffic?

rileykarson replied:
`test-fixtures` files live at https://github.com/GoogleCloudPlatform/magic-modules/tree/master/third_party/terraform/utils/test-fixtures, you can add the field there.

Traffic splitting is probably something we'll need to implement as a handwritten resource. It's configured in https://cloud.google.com/appengine/docs/admin-api/reference/rest/v1/apps.services/patch, right?

That way, users will be able to create versions in their TF config and then reference those versions from a separate traffic splitting resource.

khanali21 replied:
Added the hello-world.zip, getting excited for having this PR approved. I really appericiate your guidance and support @rileykarson.
I am looking into writing the new resource for traffic splitting.

khanali21 replied:
while running the tests locally, I am getting this error
=== CONT  TestAccAppEngineStandardAppVersion_appEngineStandardAppVersionExample
Error creating bucket : googleapi: Error 400: Empty bucket name, invalid--- FAIL: TestAccAppEngineStandardAppVersion_appEngineStandardAppVersionExample (1.25s)
    testing.go:568: Step 0 error: errors during apply:

        Error: googleapi: Error 400: Empty bucket name, invalid

          on /tmp/tf-test209836460/main.tf line 13:
          (source code not available)



        Error: error creating project test-projectz422xu7bmt (tf-test-project): googleapi: Error 400: Request contains an invalid argument., badRequest. If you received a 403 error, make sure you have the `roles/resourcemanager.projectCreator` permission

          on /tmp/tf-test209836460/main.tf line 2:
          (source code not available)


FAIL
FAIL    github.com/terraform-providers/terraform-provider-google/google 1.285s
GNUmakefile:16: recipe for target 'testacc' failed

khanali21 replied:
trying to find out how to set the bucket name variable

rileykarson replied:
It looks like there have been some commits since this was posted- are you encountering the same issue still?

modular-magician commented:
Hi!  I'm the modular magician, I work on Magic Modules.
I see that this PR has already had some downstream PRs generated. Any open downstreams are already updated to your most recent commit, 5e71b01329c87ac37f5965710c5eaaf29480c675. 
## Pull request statuses
terraform-provider-google-beta already has an open PR.
terraform-google-conversion already has an open PR.
terraform-provider-google already has an open PR.
No diff detected in Ansible.
No diff detected in Inspec.
## New Pull Requests
I didn't open any new pull requests because of this PR.

modular-magician commented:
Hi!  I'm the modular magician, I work on Magic Modules.
I see that this PR has already had some downstream PRs generated. Any open downstreams are already updated to your most recent commit, 4239c5136452dda1accea13e1257e72d25bf65d2. 
## Pull request statuses
terraform-provider-google-beta already has an open PR.
terraform-google-conversion already has an open PR.
terraform-provider-google already has an open PR.
No diff detected in Ansible.
No diff detected in Inspec.
## New Pull Requests
I didn't open any new pull requests because of this PR.

rileykarson commented on templates/terraform/examples/app_engine_standard_app_version.tf.erb:
```diff
@@ -0,0 +1,38 @@
+
+resource "google_app_engine_standard_app_version" "<%= ctx[:primary_resource_id] %>" {
+  version_id = "v1"
+  service = "<%= ctx[:vars]['service_name'] %>"
```
This value has to be `"default"`, right?

Settings the value in `vars` is used for values with variable values, particularly resource names. Since `service` must be `"default"`, it can be hardcoded.

rileykarson replied:
Running the test, I receive this error. Sounds like this is configurable, but it may be problematic to use a different service in CI?

```
------- Stdout: -------
=== RUN   TestAccAppEngineStandardAppVersion_appEngineStandardAppVersionExample
=== PAUSE TestAccAppEngineStandardAppVersion_appEngineStandardAppVersionExample
=== CONT  TestAccAppEngineStandardAppVersion_appEngineStandardAppVersionExample
--- FAIL: TestAccAppEngineStandardAppVersion_appEngineStandardAppVersionExample (29.83s)
    testing.go:568: Step 0 error: errors during apply:
        
        Error: Error creating StandardAppVersion: googleapi: Error 400: The first service (module) you upload to a new application must be the 'default' service (module). Please upload a version of the 'default' service (module) before uploading a version for the 'tf-test-servicew4qvwexmrl' service (module). See the documentation for more information. Python: (https://developers.google.com/appengine/docs/python/modules/#Python_Uploading%%20modules) Java: (https://developers.google.com/appengine/docs/java/modules/#Java_Uploading%%20modules)
        
          on /opt/teamcity-agent/temp/buildTmp/tf-test140929662/main.tf line 23:
          (source code not available)
        
        
FAIL
```

khanali21 replied:
service is configurable for the resource, I have tested locally that if a non-existence service is provided then it is created and a new version is uploaded. But it appears that if a new project is created then we must upload a version to default service. 
I have reverted the example to use the "default" service. And will test today .

khanali21 replied:
I am trying hard to get the tests passed, not sure why they keep failing, now I am not creating a project instead I am just adding the version to service default.
```
2019/08/15 10:28:15 [DEBUG] Starting graph walk: walkPlanDestroy
--- FAIL: TestAccAppEngineStandardAppVersion_appEngineStandardAppVersionExample (47.44s)
    testing.go:568: Step 0 error: After applying this step, the plan was not empty:
        
        DIFF:
        
        UPDATE: google_app_engine_standard_app_version.version_id
          deployment.#:                  "0" => "1"
          deployment.0.files.#:          "" => "0"
          deployment.0.zip.#:            "" => "1"
          deployment.0.zip.0.source_url: "" => "https://storage.googleapis.com/appengine-static-contentk6a47sozsz/hello-world.zip"
          entrypoint.#:                  "0" => "1"
          entrypoint.0.shell:            "" => "node ./app.js"
          env_variables.port:            "" => "8080"
          handlers.#:                    "0" => "0"
          id:                            "apps/my-project-1540699643621/services/default/versions/v2" => "apps/my-project-1540699643621/services/default/versions/v2"
          libraries.#:                   "0" => "0"
          name:                          "apps/my-project-1540699643621/services/default/versions/v2" => "apps/my-project-1540699643621/services/default/versions/v2"
          project:                       "my-project-1540699643621" => "my-project-1540699643621"
          runtime:                       "nodejs10" => "nodejs10"
          runtime_api_version:           "" => ""
          service:                       "default" => "default"
          threadsafe:                    "true" => ""
          version_id:                    "v2" => "v2"
        
        
        
        STATE:
        
        google_app_engine_standard_app_version.version_id:
          ID = apps/my-project-1540699643621/services/default/versions/v2
          provider = provider.google
          name = apps/my-project-1540699643621/services/default/versions/v2
          project = my-project-1540699643621
          runtime = nodejs10
          runtime_api_version = 
          service = default
          threadsafe = true
          version_id = v2
        
          Dependencies:
            google_storage_bucket.bucket
        google_storage_bucket.bucket:
          ID = appengine-static-contentk6a47sozsz
          provider = provider.google
          bucket_policy_only = false
          force_destroy = false
          location = US
          name = appengine-static-contentk6a47sozsz
          project = my-project-1540699643621
          requester_pays = false
          self_link = https://www.googleapis.com/storage/v1/b/appengine-static-contentk6a47sozsz
          storage_class = STANDARD
          url = gs://appengine-static-contentk6a47sozsz
        google_storage_bucket_object.object:
          ID = appengine-static-contentk6a47sozsz-hello-world.zip
          provider = provider.google
          bucket = appengine-static-contentk6a47sozsz
          cache_control = 
          content_disposition = 
          content_encoding = 
          content_language = 
          content_type = application/zip
          crc32c = NOB6xA==
          detect_md5hash = M4k7M6WJTeuAuQw0qaPiDw==
          md5hash = M4k7M6WJTeuAuQw0qaPiDw==
          name = hello-world.zip
          output_name = hello-world.zip
          self_link = https://www.googleapis.com/storage/v1/b/appengine-static-contentk6a47sozsz/o/hello-world.zip
          source = ./test-fixtures/appengine/hello-world.zip
          storage_class = STANDARD
        
          Dependencies:
            google_storage_bucket.bucket
FAIL
FAIL	github.com/terraform-providers/terraform-provider-google/google	47.471s
GNUmakefile:16: recipe for target 'testacc' failed
```

khanali21 replied:
@rileykarson Any feedback on this. The log shows that all the resources are properly deleted. The service version GET call return 404 which means that version is deleted and I also verified it from console, both bucket object (hello-world.zip) and bucket also get deleted.
So I am not sure why the plan is not empty.

rileykarson replied:
Terraform will fail a test step if there's a non-empty plan after applying a config. We're getting one because some fields are not returned in response bodies. Upon seeing no value for a field in the response, Terraform will unset them (as you can see in plan)

We're performing a Create with the following request:

```
{
    "deployment": {
        "zip": {
            "sourceUrl": "https://storage.googleapis.com/appengine-static-contentswmz4pxf12/hello-world.zip"
        }
    },
    "entrypoint": {
        "shell": "node ./app.js"
    },
    "envVariables": {
        "port": "8080"
    },
    "id": "v2",
    "runtime": "nodejs10"
}
```

But getting back this:

```
{
    "createdBy": "{{serviceAccount}}",
    "createTime": "2019-08-16T16:42:36Z",
    "diskUsageBytes": "3373",
    "env": "standard",
    "id": "v2",
    "instanceClass": "F1",
    "name": "apps/{{project}}/services/default/versions/v2",
    "network": {
    },
    "runtime": "nodejs10",
    "runtimeChannel": "default",
    "servingStatus": "SERVING",
    "threadsafe": true,
    "versionUrl": "https://v2-dot-{{project}}.appspot.com"
}
```

As you can see in the Terraform diff and req/response pair, `deployment`, `entrypoint`, and `env_variables` aren't present.

You can set [`ignore_read`](https://github.com/GoogleCloudPlatform/magic-modules/blob/master/overrides/terraform/property_override.rb#L46-L48) on those fields in `terraform.yaml` so that Terraform doesn't attempt to set them on reads.

khanali21 replied:
@rileykarson Once again thank you very much, I have added the ignore_read flag for deployment and entrypoint fields and for the first time I got integration tests passing horrah. :)
root@dadce36c0bce:/mm# ./integration_tests.sh
==> Checking source code against gofmt...
==> Checking that code complies with gofmt requirements...
TF_ACC=1 TF_SCHEMA_PANIC_ON_ERROR=1 go test ./google -v -run=TestAccAppEngineStandardAppVersion_appEngineStandardAppVersionExample -timeout 240m -ldflags="-X=
github.com/terraform-providers/terraform-provider-google/version.ProviderVersion=acc"
=== RUN   TestAccAppEngineStandardAppVersion_appEngineStandardAppVersionExample
=== PAUSE TestAccAppEngineStandardAppVersion_appEngineStandardAppVersionExample
=== CONT  TestAccAppEngineStandardAppVersion_appEngineStandardAppVersionExample
--- PASS: TestAccAppEngineStandardAppVersion_appEngineStandardAppVersionExample (48.42s)
PASS
ok      github.com/terraform-providers/terraform-provider-google/google 48.435s

modular-magician commented:
Hi!  I'm the modular magician, I work on Magic Modules.
I see that this PR has already had some downstream PRs generated. Any open downstreams are already updated to your most recent commit, 582c96bb53a70bf3e661f0bef19b159cd85c5bc3. 
## Pull request statuses
terraform-provider-google-beta already has an open PR.
terraform-google-conversion already has an open PR.
terraform-provider-google already has an open PR.
No diff detected in Ansible.
No diff detected in Inspec.
## New Pull Requests
I didn't open any new pull requests because of this PR.

modular-magician commented:
Hi!  I'm the modular magician, I work on Magic Modules.
I see that this PR has already had some downstream PRs generated. Any open downstreams are already updated to your most recent commit, 373c5d2f3098ca7a9c75e458ee087fcabeaaf521. 
## Pull request statuses
terraform-provider-google-beta already has an open PR.
terraform-google-conversion already has an open PR.
terraform-provider-google already has an open PR.
No diff detected in Ansible.
No diff detected in Inspec.
## New Pull Requests
I didn't open any new pull requests because of this PR.

modular-magician commented:
Hi!  I'm the modular magician, I work on Magic Modules.
I see that this PR has already had some downstream PRs generated. Any open downstreams are already updated to your most recent commit, 7988184a46c61f9a07dac9c1062baf433451faf4. 
## Pull request statuses
terraform-provider-google-beta already has an open PR.
terraform-google-conversion already has an open PR.
terraform-provider-google already has an open PR.
No diff detected in Ansible.
No diff detected in Inspec.
## New Pull Requests
I didn't open any new pull requests because of this PR.

modular-magician commented:
Hi!  I'm the modular magician, I work on Magic Modules.
I see that this PR has already had some downstream PRs generated. Any open downstreams are already updated to your most recent commit, e701ac4593ea861167e3611a5e46cb803e1e2a93. 
## Pull request statuses
terraform-provider-google-beta already has an open PR.
terraform-google-conversion already has an open PR.
terraform-provider-google already has an open PR.
No diff detected in Ansible.
No diff detected in Inspec.
## New Pull Requests
I didn't open any new pull requests because of this PR.

modular-magician commented:
Hi!  I'm the modular magician, I work on Magic Modules.
I see that this PR has already had some downstream PRs generated. Any open downstreams are already updated to your most recent commit, e701ac4593ea861167e3611a5e46cb803e1e2a93. 
## Pull request statuses
terraform-provider-google-beta already has an open PR.
terraform-google-conversion already has an open PR.
terraform-provider-google already has an open PR.
No diff detected in Ansible.
No diff detected in Inspec.
## New Pull Requests
I didn't open any new pull requests because of this PR.

khanali21 commented:
Hi @rileykarson , I have also written resource resourceAppEngineApplicationVersionSplitTraffic under 
magic-modules\src\github.com\terraform-providers\terraform-provider-google\google\resource_app_engine_application_version_split_traffic.go
and associated tests.
I am intending to create a PR for this in
https://github.com/terraform-providers/terraform-provider-google

It is a single patch request for updating the split traffic using the resource appengine_application_version_split_traffic.

Please advise.
thanks
Ali

modular-magician commented:
Hi!  I'm the modular magician, I work on Magic Modules.
I see that this PR has already had some downstream PRs generated. Any open downstreams are already updated to your most recent commit, 1cdf6cf21fb64a1a942a977a94b13484b7d7391b. 
## Pull request statuses
terraform-provider-google-beta already has an open PR.
terraform-google-conversion already has an open PR.
terraform-provider-google already has an open PR.
No diff detected in Ansible.
No diff detected in Inspec.
## New Pull Requests
I didn't open any new pull requests because of this PR.

rileykarson commented on products/appengine/terraform.yaml:
```diff
@@ -24,7 +24,36 @@ overrides: !ruby/object:Overrides::ResourceOverrides
+  StandardAppVersion: !ruby/object:Overrides::Terraform::ResourceOverride
+    id_format: "apps/{{project}}/services/{{service}}/versions/{{version_id}}"
+    import_format: ["apps/{{project}}/services/{{service}}/versions/{{version_id}}"]
+    parameters:
```
```suggestion
    mutex: "apps/{{project}}/services/{{service}}"
    parameters:
```

Only 1 operation can happen on a service at a time, so we'll need to guard our service with a mutex.

khanali21 replied:
Thank you @rileykarson.

khanali21 replied:
Should I grant you access to my fork?

rileykarson commented:
Alright, merged #2213. If you rebase off/merge `master` you should be able to create the virtual field `noop_on_destroy` and then use a [`custom_delete`](https://github.com/GoogleCloudPlatform/magic-modules/tree/master/templates/terraform/custom_delete) to skip deleting the resource if the field is set.

khanali21 commented:
@rileykarson Great stuff. Thanks for quick response. I have made the required changes. The noop_on_destroy is working as expected. I have tested with two scenarios one with noop_on_destroy true and other with false. And it works as expected.

The issue I am facing is when running the integration tests I get the following error
```
--- FAIL: TestAccAppEngineStandardAppVersion_appEngineStandardAppVersionExample (52.31s)
    testing.go:568: Step 1 error: ImportStateVerify attributes not equivalent. Difference is shown below. Top is actual, bottom is expected.

        (map[string]string) (len=1) {
         (string) (len=15) "noop_on_destroy": (string) (len=5) "false"
        }


        (map[string]string) (len=1) {
         (string) (len=15) "noop_on_destroy": (string) (len=4) "true"
        }

    testing.go:629: Error destroying resource! WARNING: Dangling resources
        may exist. The full state and error is shown below.

        Error: Check failed: AppEngineStandardAppVersion still exists at https://appengine.googleapis.com/v1/apps/cio-strat-svc-acct-mgmt-238520/services/default/versions/v2

        State: <no state>
FAIL
FAIL    github.com/terraform-providers/terraform-provider-google/google 52.341s
GNUmakefile:16: recipe for target 'testacc' failed
make: *** [testacc] Error 1
```

modular-magician commented:
Hi!  I'm the modular magician, I work on Magic Modules.
I see that this PR has already had some downstream PRs generated. Any open downstreams are already updated to your most recent commit, dc4b915977cde1f744d7a82bda0c13bd1c649c4c. 
## Pull request statuses
terraform-provider-google-beta already has an open PR.
terraform-google-conversion already has an open PR.
terraform-provider-google already has an open PR.
No diff detected in Ansible.
No diff detected in Inspec.
## New Pull Requests
I didn't open any new pull requests because of this PR.

rileykarson commented on templates/terraform/custom_delete/noop_on_destroy_appengine_version.go.erb:
```diff
@@ -0,0 +1,48 @@
+if d.Get("noop_on_destroy") == false {	
```
nit: It's a little clearer if you return immediately when the value is true instead of entering a long block when the value is false.

rileykarson commented on products/appengine/terraform.yaml:
```diff
@@ -24,7 +24,44 @@ overrides: !ruby/object:Overrides::ResourceOverrides
+    examples:
+      - !ruby/object:Provider::Terraform::Examples
+        name: "app_engine_standard_app_version"
+        primary_resource_id: "version_id"
```
```suggestion
        primary_resource_id: "version_id"
        ignore_read_extra:
          - "noop_on_destroy"
```

This will cause the `ImportStateVerify` error to go away by ignoring the field on import.

khanali21 commented:
Looks like we are getting close to merge.... I am excited. 
I will have couple of more PRs ready too once you have it merged.
```
==> Checking that code complies with gofmt requirements...
TF_ACC=1 TF_SCHEMA_PANIC_ON_ERROR=1 go test ./google -v -run=TestAccAppEngineStandardAppVersion_appEngineStandardAppVersionExample -timeout 240m -ldflags="-X=github.com/terraform-providers/terraform-provider-google/version.ProviderVersion=acc"
=== RUN   TestAccAppEngineStandardAppVersion_appEngineStandardAppVersionExample
=== PAUSE TestAccAppEngineStandardAppVersion_appEngineStandardAppVersionExample
=== CONT  TestAccAppEngineStandardAppVersion_appEngineStandardAppVersionExample
--- PASS: TestAccAppEngineStandardAppVersion_appEngineStandardAppVersionExample (52.17s)
PASS
ok      github.com/terraform-providers/terraform-provider-google/google 52.199s
```

modular-magician commented:
Hi!  I'm the modular magician, I work on Magic Modules.
I see that this PR has already had some downstream PRs generated. Any open downstreams are already updated to your most recent commit, 20ca8b5753857201ba96cb17aebc0ca5cfc00002. 
## Pull request statuses
terraform-provider-google-beta already has an open PR.
terraform-google-conversion already has an open PR.
terraform-provider-google already has an open PR.
No diff detected in Ansible.
No diff detected in Inspec.
## New Pull Requests
I didn't open any new pull requests because of this PR.

khanali21 commented:
@rileykarson Done.

modular-magician commented:
Hi!  I'm the modular magician, I work on Magic Modules.
I see that this PR has already had some downstream PRs generated. Any open downstreams are already updated to your most recent commit, 71c3e7de0273648e9d560e7aa34b5861b25e4f75. 
## Pull request statuses
terraform-provider-google-beta already has an open PR.
terraform-provider-google already has an open PR.
No diff detected in Ansible.
## New Pull Requests
I built this PR into one or more new PRs on other repositories, and when those are closed, this PR will also be merged and closed.
depends: https://github.com/GoogleCloudPlatform/terraform-google-conversion/pull/177 
depends: https://github.com/modular-magician/inspec-gcp/pull/189

rileykarson commented:
Hmm, that didn't seem to take. Can you confirm whether you added your email to your corp CLA or changed your GH settings? And if you changed your GH settings, can you confirm that your corporate email has the `primary` tag on https://github.com/settings/emails?

khanali21 commented:
I set the primary address in github to be my corporate address.

khanali21 commented:
@rileykarson I have squashed and merged in one commit, hopefully this will work.
